addons/binding/org.openhab.binding.icloud/src/main/java/org/openhab/binding/icloud/handler/BridgeHandler.java

```diff
@@ -0,0 +1,191 @@
+        iCloudDeviceHandlers.remove(device);
+    }
+
+    @Override
+    public void dispose() {
+        if (refreshJob != null) {
+            refreshJob.cancel(true);
+        }
+        super.dispose();
+    }
+
+    @Override
+    public void childHandlerInitialized(@NonNull ThingHandler childHandler, @NonNull Thing childThing) {
+        super.childHandlerInitialized(childHandler, childThing);
+        if (iCloudData != null) {
+            ((DeviceHandler) childHandler).update(iCloudData.getContent());
+        }
+    }
+
+    public void findMyDevice(String deviceId) {
+        try {
+            connection.findMyDevice(deviceId);
+        } catch (Exception e) {
+            logger.warn("Unable to execute \"find my device\" ", e);
+        }
+    }
+
+    private void startHandler() {
+        logger.debug("iCloud bridge starting handler ...");
+        config = getConfigAs(AccountThingConfiguration.class);
+
+        registerDeviceDiscoveryService();
+
+        refreshJob = scheduler.scheduleWithFixedDelay(() -> {
+            refreshData();
+        }, 0, config.refreshTimeInMinutes, TimeUnit.MINUTES);
+        logger.debug("iCloud bridge handler started.");
+    }
+
+    private void refreshDeviceData(Command command) {
+        if (command == OnOffType.ON) {
+            refreshData();
+            updateState(REFRESH, OnOffType.OFF);
+        }
+    }
+
+    private void refreshData() {
+        synchronized (synchronizeRefresh) {
+            try {
+                logger.debug("iCloud bridge refreshing data ...");
+
+                connection = new Connection(config.appleId, config.password);
+                String json = connection.requestDeviceStatusJSON();
+
+                logger.trace("json: {}", json);
+
+                iCloudData = deviceInformationParser.parse(json);
+
+                int statusCode = Integer.parseUnsignedInt(iCloudData.getStatusCode());
+                if (statusCode == 200) {
+                    updateStatus(ThingStatus.ONLINE);
+
+                    updateBridgeChannels(iCloudData);
+                    updateDevices(iCloudData.getContent());
+
+                    discoveryService.discover(iCloudData.getContent());
+                } else {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                            "Status = " + statusCode + ", Response = " + json);
+                }
+
+                logger.debug("iCloud bridge data refresh complete.");
+            } catch (Exception e) {
+                logger.warn("Unable to read data from iCloud", e);
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            }
+        }
+    }
+
+    private void registerDeviceDiscoveryService() {
+        discoveryService = new DeviceDiscovery(this);
+        service = bundleContext.registerService(DiscoveryService.class.getName(), discoveryService,

```

Comment:
Please note that bundleContext is deprecated and must not be used. Instead, you should create and register the discovery service in the handler factory and avoid any direct dependency from the handler on the discovery (while discovery on handler is ok). See e.g. how it is done in the Hue binding: https://github.com/eclipse/smarthome/blob/master/extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueThingHandlerFactory.java#L91-L92

Reply:
I did a refactoring; but not sure if this is ok yet - as I´ve not introduced an interface but used the device discovery class in the handler. Is this ok, or do I need an interface with the "discover(...)" method to decouple?

Reply:
I think it would be much cleaner if you could extract an interface that could then be defined in the same package as the handler. This would indeed cut any dependency from the handler code to the discovery package.